fieldpath/set.go

```diff
@@ -206,6 +206,45 @@ func (s *Set) WithPrefix(pe PathElement) *Set {
 	return subset
 }
 
+// Leaves returns a set containing only the leaf paths
+// of a set.
+func (s *Set) Leaves() *Set {
+	leaves := PathElementSet{}
+	im := 0
+	ic := 0
+	prevDiff := 0
+
+	// any members that are not also children are leaves
+	for im < len(s.Members.members) {
+		member := s.Members.members[im]
+		if ic >= len(s.Children.members) {
+			leaves.members = append(leaves.members, member)
+			im++
+			continue
+		}
+
+		diff := member.Compare(s.Children.members[ic].pathElement)
+		if diff-prevDiff != 0 {
```

Comment:
This subtraction is really confusing to me, I guess I expected something like:

```go
outer:
for im < len(s.Members.members) {
  member := s.Members.members[im]

  for ic < len(s.Children.members) {
    d := member.Compare(s.Children.members[ic].pathElement)
    if d == 0 {
      ic++
      im++
      continue outer
    } else if d < 0 {
      break
    } else /* if d > 0 */ {
      ic++
    }
  }
  leaves.members = append(leaves.members, member)
}
```

Reply:
Ya, that's way cleaner